### DISCUSSION TITLE
Porting to the "Cheap yellow display" (ESP32-2432S028R)

### DISCUSSION THREAD

**Discussion by witnessmenow** [harbaum/galagino#18] [General]
Hello, awesome project, I just found it today!

I've been using similar hardware at the moment that you might find interesting as a good starting point or building block for this project. 

It's the "ESP32-2432S028R", but since there is a bunch of similarly looking and named displays, I decided it might be a nice idea to give it a rebranding and try build a bit of a community around it, so I am renaming it the [Cheap Yellow Display](https://github.com/witnessmenow/ESP32-Cheap-Yellow-Display)

It's an ESP32 board with a built in 320x240 ILI9341 and only costs about $15 delivered on Aliexpress. It also even has a SC8002B included too (although no speaker, just a JST connector for one). So really similar hardware to what the current project uses.

The only downside is the GPIO. I haven't tested it fully yet but from looking at 
- It only has 4 GPIO broken out to JST connectors (35,22,21 and 27), 
- TX and RX are also broken out to JST so while not ideal, they are potentially usable as GPIO
- There is a boot button (GPIO 0) that could be used as a less frequently used button like coin or start

It should also be possible to use an I2C expander like the PCF8575, it will require some code changes but I think it wouldn't be too big of a deal.

Over the next few weeks, I'm going to try get this project up and running on the Cheap yellow display, once I've had a chance to try out some of the different input methods

Possibly a different discussion topic, but I also see in the docs that it would be possible to extract the roms on the ESP32, would there be much work involved in doing something like this? I'm not sure if you've seen it before but the [ESP web tools](https://esphome.github.io/esp-web-tools/) are a pretty cool way of providing a compiled binary of a project that end users can just flash directly from their browser. But the current structure of this project involving the roms at compile time, means that any binaries provided would include the roms, which is probably a problem. A solution where the roms were loaded from SD card or something could provide a path to allowing people to webflash the project.

**Comment by harbaum** [comment]
The current Galagino also includes the Z80 emulation with an incompatible license. I tried to reach out to Marat but got no reply. So currently the ROMs aren't the only part preventing redistribution in binary form.

Secondly, the ROMs aren't just included 1:1. Instead the video and audio ROMs are preprocessed by those python scripts. This processing would have to be moved to the ESP32 as well. This in turn would mean that ROMs are loaded into ram and ram is pretty limited on the esp32. And yes, there are esp32 with huge external rams (those wrover boards) but the external ram is too slow to run Galagino. 

It might be possible to do all this. But I am not sure it's worth the hassle. The current setup works fine and it's just a matter of running a few python scripts to get everything up and running.

**Reply by witnessmenow** [reply]
Cool, good to know the full challenges with it! Yeah the python solution looks good

I'm looking forward to trying to get something up and running with the inputs so I can give if a go!

**Comment by witnessmenow** [comment]
I got it working on the Cheap Yellow display today. I decided to use a nunchuck as the input option as they are so cheap and can easily be used with the JST to header cable that comes with the display. Seems to work well

![image](https://github.com/harbaum/galagino/assets/1562562/e20f2eab-5d37-4045-a7a1-e2e5e1af2c14)

![image](https://github.com/harbaum/galagino/assets/1562562/2cb292c1-639f-4353-8202-c0521f9c1604)

Changes made:

- Added a conv_win.sh is executable in windows powershell (I just added "python" to the start of each command so it's probably also executable in unix and mac, but not tested)
- Made some changes in video.cpp to allow for non standard SPI pins and no reset pin
- Added a `CHEAP_YELLOW_DISPLAY_CONF` define in the config.h
- Added support for Nintendo Wii Nunchucks, can be enabled in config.h

~~Code is here: https://github.com/BLough-Heads/galagino~~
EDIT: update repo here: https://github.com/witnessmenow/galagino

If your interested in a PR, I'm more than happy to raise one, but I would like to do some more testing first. 

Do you have a style guide or code formater you are using @harbaum ? There actually isn't that much code changes in this, but my auto formater made a tonne of changes to it. I can re-apply the changes again without the formater if needed.

**Reply by harbaum** [reply]
Sounds great. And yes, a PR is very welcome.

I don't have a special code style. Actually the Arduino IDE, Emacs and Github somewhat disagree on their style, so it's a mess, anyway.

I'd very much prefer a PR that does not include too many style changes as that makes the diffs/patches basically unreadable.

**Reply by witnessmenow** [reply]
Great!

No worries, I'll make the code changes again and disable the formatter

**Reply by witnessmenow** [reply]
https://github.com/harbaum/galagino/compare/main...witnessmenow:galagino:main

Here is a clean version of the changes with no formatting changes applied @harbaum , I'll raise a PR after I've tested audio (which will require changes as it uses pin26, not 25)

**Reply by witnessmenow** [reply]
Audio works with my latest changes. 

I haven't done the mod yet and it is a bit loud. Adding a pot to control the gain might be the best thing.

I'll do some more testing over the next few days and raise the PR

**Reply by harbaum** [reply]
Excellent. Thanks a lot. I'll order one of the yellow boards to give it a try as well.

**Reply by witnessmenow** [reply]
No worries at all, thanks so much for the project!

If you have any issues or questions let me know.

Will I add some documentation on the Readme for the Cheap Yellow Display and maybe separately the nunchuck? Completely up to you, I don't mind documenting it somewhere else if that makes more sense either.

**Reply by harbaum** [reply]
Great. And yes, a short prominent note that the cheap yellow boards are supported is welcome. 

And having the nunchuck part explained in more detail on a separate page also sounds perfect.

**Reply by witnessmenow** [reply]
I've opened the PR there. Please feel free to ask for changes if there is something you would prefer done differently!

**Reply by Rotorhead99** [reply]
Hi guys, 

great work you have done so far. I have a CYD and galagino ist running well, nunchuk fine, but a can´t get audio running. The "7-HelloRadio.ino" from witnessmenow is running fine, but no sound from galagino... Any ideas? 

Best regards, 

Dirk

**Reply by witnessmenow** [reply]
Hi @Rotorhead99 , I've raised a PR to fix the audio on the CYD

https://github.com/harbaum/galagino/pull/33

It's a fairly simple change you can add to galagino.ino to fix the issue

Also make sure to use ESP32 Core 2.0.9 as per the readme as there is an issue in 2.0.10 and 2.0.11 (seems like it will be fixed in 2.0.12)

**Reply by donniebrasc** [reply]
But would be possible to run a txt. editor and store to sd with a interface? DuckHunter? #Dr.Evil#SinisterLOL

**Comment by harbaum** [comment]
I don't see a speaker in your photos. Is audio working?

**Reply by witnessmenow** [reply]
Haven't got a chance to test if yet, will do it over the next few days if I get time.

Some people on my discord have used a speaker with the display in other sketches, but apparently out of the box the gain on the amp is very high, so it kind of requires a hardware mod to make it sound reasonable (I haven't tried any of it)

https://github.com/witnessmenow/ESP32-Cheap-Yellow-Display/tree/main/Mods#audio-amp-gain-mod

**Comment by harbaum** [comment]
The nunchuck is connected via i2c, right? It could actually be used for a second player on the cabinet. 

I also considered using some Bluetooth or WLAN control method. But I'd expect this to consume significant CPU cycles now and then, and I am afraid this would interfere with the fluid gameplay.

**Reply by witnessmenow** [reply]
Yup i2c so would be a good candidate.

Yeah I think the delay might be an issue too. I was even worried the polling of the nunchuck for the input via i2c might impact, but it seems fine!

It is also possible to have two nunchucks working.
The nunchuck has a fixed address so they can't be on the same bus, but they can be configured on two different busses

Just as an fyi, the way you did the button press bit mask made it really easy to swap in the nunchuck code, I was so happy to see it done that way as I originally had visions of changing button presses In a Million different places 😅

**Comment by harbaum** [comment]
My CYD arrived and the display was messed up (rotated by 90 deg, "running" image and wrong colors).

Turned out it needed a slightly different setup of the Memory Access Control register. There's now a config option to set custom values for that register.

It seems there's a reason these displays are so cheap. They have a very low contrast. The digitizer/touchsceen overlay may contribute to this. But the display contrast is very low and black is a medium grey and the colors aren't very vibrant. The slightly more expensive displays advertised as "IPS" look way better. Much brighter, way more contrast and more saturated colors. 

I still do have some bare IPS displays. I think i'll try to replace the one on my CYD.

**Comment by harbaum** [comment]
The CYD is on the left. In real life the difference looks worse as the camera seems to adopt to the different brightness. In real life the CYD black looks grey.

![IMG_20230706_203659.jpg](https://github.com/harbaum/galagino/assets/9105084/4cc49844-2744-447d-b274-5a882bd2e2ab)

**Reply by witnessmenow** [reply]
Oh that's a pity!

I don't currently have any better displays here to compare it against, but I have 3 CYD here and just comparing my ones to your pictures, the colours are definitely closer to the right screen than the left. (Not a perfect test, but its something!)

![image](https://github.com/harbaum/galagino/assets/1562562/5c9baa2e-6f66-4fb5-bbc1-868824d3e731)

I'm also curious why your one needed the different Memory Access Control register. One of the the big ideas behind the CYD being a nice platform for people is it being a common piece of hardware. Did you buy it from one of my links? Could you let me know which one and I'll pick one up to compare. Maybe they are selling basically knockoffs I need to remove that store link, or maybe its the luck of the draw!

**Reply by harbaum** [reply]
I bought this one:
https://de.aliexpress.com/item/1005005262421075.html

**Reply by harbaum** [reply]
Interestingly, mine has an additional USB-C port, which seems to be simply wired in parallel with the micro USB.

**Comment by anthonyjclarke** [comment]
Hi

I had this issue. Fix and lots of good support on this Discord channel https://discord.com/channels/630078152038809611/1109228361441620028/1151315893368930415

Basically drop ESP32 board library to 2.09 and it worked (for me)

What version are you using?

**Comment by arminth** [comment]
I have worked on a version to support both the NUNCHUK and the Nintendo Classic Controller. The latter provides all Buttons needed including coin, start and home (which I have added to the emulation.h). At the moment, it works fine and to switch from NUNCHUK to CLASSIC, I need to change the config.h. I will continue to work on this and come up with a autodetection of the Nintendo-controllers suitable. Still work in progress but looks promising already.
Also remixed new Plywood case in 4mm Plywood (that is, what we have available at the local hardware store) and added the STLs and OpenSCAD-file for the Marquee and Controll-Part.
It is 30% bigger that the original ones, but I like it and it fit's great with the CYD.
When the PCBs arrive for the WII-plug-sniffers, I will add a mount for this as well so you can plug the Nunchuk or classic controller to your liking. I'll keep you posted on the progress.
Cheers
Armin
![IMG_20240205_150603059](https://github.com/harbaum/galagino/assets/16935720/0599b919-e1e8-4476-b3d4-f9bda824f39b)

**Comment by arminth** [comment]
Ok guys,
I have done it! my first pull-request!
I reworked the code to enable both Nunchuk and classic controllers. I also enhanced the functionality of Nunchuk to include COIN (C-Button), START (press C and Z together) and emulatio-reset (Press Z, C and Joystick down at once).
Classic controller uses DPAD, Select = COIN, Start=Start, Home=emulation-reset and a=FIRE.
I also finished the SVG-files for the 4mm plywood cabinet and the STLs. I will upload to Thingiverse and Printables and post the links here. I added a 3D-printed mount for @witnessmenow  Brian Lough's Nunchuck-PCB with the notch an the 2 resistors (great work btw!). With this screwed to the backcover, you can easily switch between controllers.

I hope, you like my changes. I love it, as there is no hardwarebutton needed anymore!

Cheers
Armin

**Comment by arminth** [comment]
https://www.printables.com/de/model/765208-galagino-arcade-case-4mm-plywood-for-cyd-and-ninte
https://www.thingiverse.com/thing:6485594

**Reply by arminth** [reply]
Thingiverse messes up SVG.... Printables doesn't accept Lightburn Projects! Grrrrrr!

**Comment by arminth** [comment]
I have appended my pull request to autodetection of type of Nintendo-controller and added some functionality to classsic controller (use DPAD, left joystick or both)

**Comment by retronicdesign** [comment]
Hey there! very nice project!
I build one myself and added the PCF8575 on the I2C for the joystick/buttons.
Everything works fine except the audio pitch is too high. It was working properly on the ESP32 board but now it's different ont the cheap yellow display.  I noticed that sound samples are playing at the right pitch. Only "waves" are too high.
Any suggestion?
Thank you!
